### PR TITLE
REP-4225 Bypass URL

### PR DESCRIPTION
**Pull request by adrianjgeorge**

### CODE REVIEW

**Comment by adrianjgeorge** [comment]
Just the config so far

**Comment by dmnjohns** [comment]
Config looks good. 👍

**Comment by adrianjgeorge** [comment]
track the build, https://jenkins.openrepose.org/job/repose-gradle-pull_request/614/

File: repose-aggregator/core/repose-core/src/main/java/org/openrepose/powerfilter/PowerFilterChain.java [hunk 1]
```diff
@@ -78,16 +79,19 @@
...
     private boolean filterChainAvailable;
     private TimerByCategory filterTimer;
+    private Optional<String> bypassUrl;
 
     public PowerFilterChain(List<FilterContext> filterChainCopy,
                             FilterChain containerFilterChain,
                             PowerFilterRouter router,
-                            MetricsService metricsService)
+                            MetricsService metricsService,
+                            Optional<String> bypassUrl)
```
Comment (dmnjohns) [inline repose-aggregator/core/repose-core/src/main/java/org/openrepose/powerfilter/PowerFilterChain.java]: Oddly, IntelliJ warns on using the `Optional` type for a field or property. I wonder why...

File: repose-aggregator/core/repose-core/src/main/java/org/openrepose/powerfilter/PowerFilterChain.java [hunk 2]
```diff
@@ -143,9 +147,13 @@ private void splitRequestHeaders(HttpServletRequestWrapper request) {
      */
     private List<FilterContext> getFilterChainForRequest(String uri) {
         List<FilterContext> filters = new LinkedList<>();
-        for (FilterContext filter : filterChainCopy) {
-            if (filter.getUriPattern() == null || filter.getUriPattern().matcher(uri).matches()) {
-                filters.add(filter);
+        if (bypassUrl.isPresent() && Pattern.compile(bypassUrl.get()).matcher(uri).matches()) {
```
Comment (dmnjohns) [inline repose-aggregator/core/repose-core/src/main/java/org/openrepose/powerfilter/PowerFilterChain.java]: You could optionally replace the logical and operator here with more calls on the Optional.
`bypassUrl.map(url -> Pattern.compile(url).matcher(uri).matches()).orElse(false)`

File: repose-aggregator/core/repose-core/src/main/java/org/openrepose/powerfilter/PowerFilter.java [hunk 3]
```diff
@@ -334,7 +334,10 @@ private PowerFilterChain getRequestFilterChain(HttpServletResponse httpResponse,
                 //Update the JMX bean with our status
                 configurationInformation.updateNodeStatus(clusterId, nodeId, false);
             } else {
-                requestFilterChain = new PowerFilterChain(filterChain, chain, router, metricsService);
+                requestFilterChain = new PowerFilterChain(filterChain, chain, router, metricsService,
+                        Optional.of(currentSystemModel.get().getReposeCluster().stream()
+                                .filter(cluster -> cluster.getId().equals(clusterId)).findFirst()
+                                .get().getFilters().getBypassUriRegex()));
```
Comment (dmnjohns) [inline repose-aggregator/core/repose-core/src/main/java/org/openrepose/powerfilter/PowerFilter.java]: You were right -- this is functional and concise, but it really isn't very pretty.

**Comment by dmnjohns** [comment]
I left some comments, but as per usual, they are more notes to consider than actions to take.

Overall, it looks good to me. 👍

File: repose-aggregator/core/repose-core/src/main/java/org/openrepose/powerfilter/PowerFilterChain.java (hunk 2 above)
Comment (adrianjgeorge) [inline repose-aggregator/core/repose-core/src/main/java/org/openrepose/powerfilter/PowerFilterChain.java]: Does the java optional have a map? i didn't see it when i looked through the list.

Comment (dmnjohns) [inline repose-aggregator/core/repose-core/src/main/java/org/openrepose/powerfilter/PowerFilterChain.java]: Sure does.

https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html#map-java.util.function.Function-

Comment (adrianjgeorge) [inline repose-aggregator/core/repose-core/src/main/java/org/openrepose/powerfilter/PowerFilterChain.java]: Huh, it does wonder why that didn't show up in the ide as an option. I'll switch it.

Comment (wdschei) [inline repose-aggregator/core/repose-core/src/main/java/org/openrepose/powerfilter/PowerFilterChain.java]: Yep.
 - https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html#map-java.util.function.Function-

File: repose-aggregator/core/repose-core/src/main/java/org/openrepose/powerfilter/PowerFilter.java [hunk 4]
```diff
@@ -334,7 +334,10 @@ private PowerFilterChain getRequestFilterChain(HttpServletResponse httpResponse,
                 //Update the JMX bean with our status
                 configurationInformation.updateNodeStatus(clusterId, nodeId, false);
             } else {
-                requestFilterChain = new PowerFilterChain(filterChain, chain, router, metricsService);
+                requestFilterChain = new PowerFilterChain(filterChain, chain, router, metricsService,
+                        Optional.ofNullable(currentSystemModel.get().getReposeCluster().stream()
+                                .filter(cluster -> cluster.getId().equals(clusterId)).findFirst()
+                                .get().getFilters().getBypassUriRegex()));
```
Comment (wdschei) [inline repose-aggregator/core/repose-core/src/main/java/org/openrepose/powerfilter/PowerFilter.java]: Wow! That looks ugly.

Comment (adrianjgeorge) [inline repose-aggregator/core/repose-core/src/main/java/org/openrepose/powerfilter/PowerFilter.java]: Yeah, i'm very quickly becoming not a fan of the token functional constructs they added in java 8.

File: repose-aggregator/core/repose-core/src/main/java/org/openrepose/powerfilter/PowerFilterChain.java (hunk 1 above)
Comment (wdschei) [inline repose-aggregator/core/repose-core/src/main/java/org/openrepose/powerfilter/PowerFilterChain.java]: That is crazy. It says:

> Optional was designed to provide a limited mechanism for library method return types where there needed to be a clear way to represent "no result". Using a field with type `java.util.Optional` is also problematic if the class needs to be `Serializable`, which `java.util.Optional` is not.

Comment (adrianjgeorge) [inline repose-aggregator/core/repose-core/src/main/java/org/openrepose/powerfilter/PowerFilterChain.java]: Thankfully, i don't care is it's serializable or not.

**Comment by wdschei** [comment]
Retest this please.

**Comment by wdschei** [comment]
This was the last run:

- https://jenkins.openrepose.org/job/repose-gradle-pull_request/616/

All 8 failures were `java.util.concurrent.TimeoutException` at `framework.ReposeValveTest.waitUntilReadyToServiceRequests(ReposeValveTest.groovy:121)`.

**Comment by wdschei** [comment]
Current run is:

- https://jenkins.openrepose.org/job/repose-gradle-pull_request/617/

**Pull request merged**